lib/travis/build/script/elm.rb

```diff
@@ -0,0 +1,183 @@
+        end
+
+        def announce
+          super
+          sh.cmd 'elm --version', echo: true
+          sh.cmd 'elm-test --version', echo: true
+
+          # elm-format doesn't have --version,
+          # but the first line of `elm-format --help` prints the version
+          sh.cmd 'elm-format --help | head -n 1', echo: true
+        end
+
+        def setup
+          super
+
+          sh.echo 'Elm for Travis-CI is not officially supported, ' \
+          'but is community maintained.', ansi: :green
+          sh.echo 'Please file any issues using the following link',
+            ansi: :green
+          sh.echo '  https://github.com/travis-ci/travis-ci/issues' \
+            '/new?labels=community:elm', ansi: :green
+          sh.echo 'and mention \`@avh4\`, \`@lukewestby\`, \`@stoeffel\` and \`@rtfeldman\`' \
+            ' in the issue', ansi: :green
+
+
+          sh.if '! -d sysconfcpus/bin' do
+            install_sysconfcpus
+          end
+
+          sh.fold 'install.elm' do
+            install_elm
+            install_elm_test
+            install_elm_format
+          end
+        end
+
+        def script
+          sh.cmd 'elm-format --validate . && elm-test'
+        end
+
+        def setup_cache
+          if data.cache?(:elm)
+            sh.fold 'cache.elm' do
+              # Cache the ~/.elm directory.
+              directory_cache.add '$HOME/.cache/elm', '$HOME/.elm'
+
+              directory_cache.add '$HOME/.cache/elm-stuff', 'elm-stuff'
+
+              if elm_major_version == 0 && elm_minor_version <= 18
+                # In Elm 0.18, some put their tests in ./test instead of ./tests
+                directory_cache.add '$HOME/.cache/elm-test-stuff', 'test/elm-stuff'
+              end
+
+              # In Elm 0.18+, all tests live in tests/ by default
+              # (whereas previously tests/ was allowed, but so was test/)
+              directory_cache.add '$HOME/.cache/elm-tests-stuff', 'tests/elm-stuff'
+
+              # we build sysconfcpus from source, so cache the result
+              directory_cache.add '$HOME/.cache/sysconfcpus-cache', 'sysconfcpus'
+            end
+          end
+        end
+
+        def cache_slug
+          super << '-elm-' << elm_version
+        end
+
+        private
+
+          def elm_version
+            config[:elm].to_s
+          end
+
+          def elm_version_number(index)
+            elm_version.split(".")[index]
+          end
+
+          def elm_major_version
+            elm_version_number 0
+          end
+
+          def elm_minor_version
+            elm_version_number 1
+          end
+
+          def elm_patch_version
+            elm_version_number 2
+          end
+
+          def elm_test_version
+            config[:elm_test].to_s
+          end
+
+          def elm_format_version
+            config[:elm_format].to_s
+          end
+
+          def install_elm
+            npm_install "-g elm@elm-#{elm_version}"
```

Comment:
Did Elm's numbering scheme change? https://staging.travis-ci.org/BanzaiMan/travis_staging_test/jobs/732882#L824-L828 shows:

```
$ npm install -g elm@elm-0.19.0
npm ERR! code ETARGET
npm ERR! notarget No matching version found for elm@elm-0.19.0
npm ERR! notarget In most cases you or one of your dependencies are requesting
npm ERR! notarget a package version that doesn't exist.
```

Reply:
Dropping `elm-` works: https://staging.travis-ci.org/BanzaiMan/travis_staging_test/builds/732909

Reply:
I just noticed that the use of `npm_install` messes up the folds, as `npm_install` itself sets up a fold, and nested folds don't work well. The method should be overridden to remove the fold.

Additionally, it checks whether or not the packages are locked, and invokes different commands. This may not be appropriate for Elm. https://github.com/travis-ci/travis-build/blob/149ff67390be18b128975cc298d4ae4be3d001b2/lib/travis/build/script/node_js.rb#L190-L195

Reply:
See https://staging.travis-ci.org/BanzaiMan/travis_staging_test/jobs/732930 for an example build with the fold fixed.

Reply:
Thanks! Since these all use `npm install -g`, I made a separate method in f53ea8a36a6ee1ab463e482bcd1bdd5e1f182708 which runs that and switched everything to do that.

Also fixed the incorrect npm version strings while I was at it. 👍